crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,1179 @@
+            }
+            ModuleResolutionPathBufInner::StandardLibrary(path) => {
+                ModuleResolutionPathRefInner::StandardLibrary(path)
+            }
+            ModuleResolutionPathBufInner::SitePackages(path) => {
+                ModuleResolutionPathRefInner::SitePackages(path)
+            }
+        }
+    }
+}
+
+impl<'a> From<&'a ModuleResolutionPathBuf> for ModuleResolutionPathRef<'a> {
+    fn from(value: &'a ModuleResolutionPathBuf) -> Self {
+        ModuleResolutionPathRef(ModuleResolutionPathRefInner::from(&value.0))
+    }
+}
+
+impl<'a> PartialEq<ModuleResolutionPathBuf> for ModuleResolutionPathRef<'a> {
+    fn eq(&self, other: &ModuleResolutionPathBuf) -> bool {
+        match (self.0, &other.0) {
+            (
+                ModuleResolutionPathRefInner::Extra(self_path),
+                ModuleResolutionPathBufInner::Extra(other_path),
+            )
+            | (
+                ModuleResolutionPathRefInner::FirstParty(self_path),
+                ModuleResolutionPathBufInner::FirstParty(other_path),
+            )
+            | (
+                ModuleResolutionPathRefInner::StandardLibrary(self_path),
+                ModuleResolutionPathBufInner::StandardLibrary(other_path),
+            )
+            | (
+                ModuleResolutionPathRefInner::SitePackages(self_path),
+                ModuleResolutionPathBufInner::SitePackages(other_path),
+            ) => *self_path == **other_path,
+            _ => false,
+        }
+    }
+}
+
+impl<'a> PartialEq<FileSystemPath> for ModuleResolutionPathRef<'a> {
+    fn eq(&self, other: &FileSystemPath) -> bool {
+        self.0.as_file_system_path() == other
+    }
+}
+
+impl<'a> PartialEq<ModuleResolutionPathRef<'a>> for FileSystemPath {
+    fn eq(&self, other: &ModuleResolutionPathRef<'a>) -> bool {
+        self == other.0.as_file_system_path()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use insta::assert_debug_snapshot;
+
+    use crate::db::tests::{create_resolver_builder, TestCase, TestDb};
+    use crate::supported_py_version::SupportedPyVersion;
+
+    use super::*;
+
+    // Replace windows paths
+    static WINDOWS_PATH_FILTER: [(&str, &str); 1] = [(r"\\\\", "/")];
+
+    #[test]
+    fn constructor_rejects_non_pyi_stdlib_paths() {
+        assert!(ModuleResolutionPathBuf::standard_library("foo.py").is_none());
+        assert!(ModuleResolutionPathBuf::standard_library("foo/__init__.py").is_none());
+        assert!(ModuleResolutionPathBuf::standard_library("foo.py.pyi").is_none());
+    }
+
+    fn stdlib_path_test_case(path: &str) -> ModuleResolutionPathBuf {
+        ModuleResolutionPathBuf::standard_library(path).unwrap()
+    }
+
+    #[test]
+    fn stdlib_path_no_extension() {
+        assert_debug_snapshot!(stdlib_path_test_case("foo"), @r###"
+        ModuleResolutionPath::StandardLibrary(
+            "foo",
+        )
+        "###);
+    }
+
+    #[test]
+    fn stdlib_path_pyi_extension() {
```

Comment:
Wow, this is very intensive testing that you added for path and I really apprevciate your care and time that you took to write all of them. 

I'm feeling a bit conflicted about some of the tests. I can see how they're useful when making the implementation but I fear they'll create a larger long-term coast than the value they provide (and writing the tests also comes at a cost):

* A lot of these (not all) tests test simple trivia. I can see that having the test can help when implementing the function but I'm not sure what the long-term value of the tests are and maintaining the code comes at a cost (and just looking at 1500 lines of tests is intimiating, and hides the real important tests).
* Most of the tests are precisely testing the implementation (and often match the implementation 1:1). Any change to the implementation will require a change to the test, significantely reducing the value of the tests and increasing maintenance cost long term. The most valuable tests are tests that keep functioning even when you refactor code. I think the resolver tests are a good exmaple. We refactored the implementation 2 times but the tests are still mostly unchanged, providing a safety net when making the refactor. I don't this applies to some tests in this module. Refactoring the path representation would require rewriting all these tests.

This makes me wonder how many of these tests we still need and if we could get the same coverage by extending our module resolver tests (e.g. by porting the tests from import resolver). I think we should at least reduce the number of tests. We don't need to aim for full exhaustiveness when testing. 
For example, do we need all of the `stdlib_path_*` tests or would one where it returns `None` and one where it returns `Some` be sufficient? I do like the `module_name_*` but do we need all of them? Are all of them testing different code paths or are we trying to test for full exhaustiveness? 


We don't need to do this as part of this PR, the PR is already big enough. But increasing the module resolver coverage seems desirable to me, especially if it helps to avoid some more "boilerplate"ish tests in here.

Reply:
Thanks! I've simplified the tests and reduced the overall number quite a lot.